torch/csrc/jit/codegen/cuda/shape_expr_memo.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir.h>
+
+#include "glfdc/eval.h"
+#include "glfdc/expr_builder.h"
+
+#include <unordered_map>
+
+namespace glfdc {
+struct ExpressionBuilder;
+struct EvalState;
+} // namespace glfdc
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace kir {
+
+class ExprBuilderVisitor : IrVisitor {
+ public:
+  explicit ExprBuilderVisitor(
+      glfdc::ExpressionBuilder& bld,
+      const Kernel& kernel);
+
+  c10::optional<glfdc::Expr> build(const Val* value) {
+    auto ref = build_operand(value);
+    return builder_->create_expr(ref);
+  }
+
+  glfdc::Operand build_operand(const Val* value) {
+    TORCH_CHECK(value->isScalar());
+    TORCH_CHECK(value->dtype() == DataType::Int);
```

Comment:
nit: Use `TORCH_INTERNAL_ASSERT` if these assertions are to validate the internal logic. `TORCH_CHECK` is for validating user inputs. It should be always used with an error message that's understandable by the user.

Reply:
Sure, think I changed most of them, will be still tweeking messages for TORCH_INTERNAL_ASSERT pending review.